### PR TITLE
Validate the ingestor config file

### LINKED ISSUE #826: No validation of config in ingestor
### Expected behaviour

When user supplies mal-formed ingestor configuration I expect errors to be reported to the user, instead exception with `KeyError` is raised.

### Actual behaviour

ingestor does not check that `src_varname` is valid for source product, resulting in errors like following (reported on slack):

```
File "/root/miniconda3/envs/odc/lib/python3.6/site-packages/datacube/scripts/ingest.py", line 153, in ensure_output_type
    output_type = morph_dataset_type(source_type, config, index, storage_format)
  File "/root/miniconda3/envs/odc/lib/python3.6/site-packages/datacube/scripts/ingest.py", line 72, in morph_dataset_type
    for spec in config['measurements']]
  File "/root/miniconda3/envs/odc/lib/python3.6/site-packages/datacube/scripts/ingest.py", line 72, in <listcomp>
    for spec in config['measurements']]
KeyError: '1'
```

### Steps to reproduce the behaviour

Create ingestor config with mis-spelled `src_varname:` attribute for one of the measurements, then run ingestor.

Error is here:

https://github.com/opendatacube/datacube-core/blob/73d98a36d69a1a4aaee8cc89945cf652a6de6260/datacube/scripts/ingest.py#L71-L72

no prior validation happened, yet code assumes that `src_varname` 
(a) exists in the config (maybe that part is checked with schema?)
(b) points to valid measurement in the source product

### PR DESCRIPTION
**Pull request by jdh-ama**
### Reason for this pull request

To validate ingestor configuration files prior to processing and provide a meaningful error message to the user. 

### Proposed changes

- Create ingestor configuration schema
- Validate ingestor configuration files against the schema prior to processing using jsonschema
- Add test case to verify that the validation catches errors
- Fix some unrelated formatting issues that cause a pre-commit hook to fail preventing commits

 - [X] Closes #826
 - [X] Tests added / passed
 - [(did not update whats_new)] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

**Comment by Kirill888** [comment]
@jdh-ama this is great, thank for reverse engineering ingestor schema, I gonna make some small changes and merge it.

File: datacube/model/__init__.py [hunk 1]
```diff
@@ -574,10 +582,11 @@ def __eq__(self, other):
         if not isinstance(other, GridSpec):
             return False
 
-        return (self.crs == other.crs
-                and self.tile_size == other.tile_size
-                and self.resolution == other.resolution
-                and self.origin == other.origin)
+        crs_result = self.crs == other.crs
```
Comment (Kirill888) [inline datacube/model/__init__.py]: Can we revert this change please? Original check short circuits computation, while this one forces each one to be computed and forces creation and choosing names for temporary variables. We should probably re-order so that expensive comparison `crs==crs` is last and cheaper ones are first.

File: datacube/scripts/ingest.py [hunk 2]
```diff
@@ -408,19 +409,30 @@ def ingest_cmd(index,
...
         config = load_config_from_file(config_file)
+    elif load_tasks:
+        config, tasks = load_tasks_(load_tasks)
+    else:
+        click.echo('Must specify exactly one of --config-file, --load-tasks')
+        sys.exit(-1)
+
+    try:
+        # ignore the added filename key which is not part of the schema
+        tmp_config = {key:config[key] for key in config if key != 'filename'}
```
Comment (Kirill888) [inline datacube/scripts/ingest.py]: Is this for when going via `load_tasks_(...)`? Can we just drop it from `config`, is it being used later on? Or maybe add it into schema as an optional string parameter?

**Review by Kirill888 — CHANGES_REQUESTED** [review]
I thought I would just make changes myself, but I didn't realise it's in a fork, so it's a bit tricky. some minor changes please

File: datacube/model/__init__.py (hunk 1 above)
Comment (jdh-ama) [reply datacube/model/__init__.py]: I reverted the changes.

File: datacube/scripts/ingest.py (hunk 2 above)
Comment (jdh-ama) [reply datacube/scripts/ingest.py]: I'm hesitant to add filename to the schema because it is created and used internally by the ingestor. I think that the schema should only contain fields that a user writing an ingestor configuration file would be concerned about. That allows the schema to also be used as documentation. Instead of copying the dictionary which is O(n) complexity I have changed to temporarily deleting the filename key for validation, delete and set are O(1) operations.

**Comment by jdh-ama** [comment]
@Kirill888 You may want to double check the required fields listed in the schema. Those were a little difficult to discern just based on the existing examples.

**Comment by Kirill888** [comment]
@jdh-ama yes, "a little difficult" is probably an understatement. I'll just merge it and then we might have to fix things as we find more.

**Review by Kirill888 — APPROVED** [review]
:+1:
